source/ada/lsp-ada_handlers-refactor_change_parameter_mode.adb

```diff
@@ -0,0 +1,415 @@
+
+      Workspace_Edits : LSP.Messages.WorkspaceEdit renames Apply.params.edit;
+
+      Node : constant Ada_Node :=
+        Document.Get_Node_At (Context, Self.Where.position);
+
+      Target_Subp               : Defining_Name := No_Defining_Name;
+      Target_Parameters_Indices : constant Parameter_Indices_Range_Type :=
+        (First => Positive (Self.First_Param_Index),
+         Last  => Positive (Self.Last_Param_Index));
+
+      Changer : Mode_Changer;
+      Edits   : Laltools.Refactor.Edit_Map;
+
+      function Analysis_Units return Analysis_Unit_Array is
+        (Context.Analysis_Units);
+      --  Provides the Context Analysis_Unit_Array to the Mode_Changer
+
+      function Get_Text_Document
+        (File_URI : LSP.Types.LSP_String)
+         return LSP.Messages.VersionedTextDocumentIdentifier;
+      --  Fetches a Document_Access given the File_URI and then creates a
+      --  OptionalVersionedTextDocumentIdentifier. This function is used as
+      --  an OptionalVersionedTextDocumentIdentifier provider to the
+      --  LSP.Lal_Utils.To_Workspace_Edit, which transforms an Edit_Map into
+      --  a WorkspaceEdit.
+
+      function Value (M : LSP.Types.LSP_String) return Ada_Mode;
+      --  Converts an LSP_String to an Ada_Mode
+
+      -----------------------
+      -- Get_Text_Document --
+      -----------------------
+
+      function Get_Text_Document
+        (File_URI : LSP.Types.LSP_String)
+         return LSP.Messages.VersionedTextDocumentIdentifier
+      is
+         Target_Text_Document : constant LSP.Ada_Documents.Document_Access :=
+           Message_Handler.Get_Open_Document (File_URI);
+         use type LSP.Ada_Documents.Document_Access;
+
+      begin
+         --  If the target textDocument hasn't been opened in the editor
+         --  then ALS hasn't received an open notification before. Therefore
+         --  Target_Text_Document will be null.
+         --  In that case, its OptinalVersionedTextDocumentIdentifier will
+         --  be null.
+
+         if Target_Text_Document = null then
+            return LSP.Messages.VersionedTextDocumentIdentifier'
+              (File_URI, LSP.Messages.Nullable_Number'(Is_Set => False));
+
+         else
+            return Target_Text_Document.Versioned_Identifier;
+         end if;
+      end Get_Text_Document;
+
+      -----------
+      -- Value --
+      -----------
+
+      function Value (M : LSP.Types.LSP_String) return Ada_Mode is
+         use type LSP.Types.LSP_String;
+      begin
+         if M = "in" then
+            return Ada_Mode_In;
+         elsif M = "out" then
+            return Ada_Mode_Out;
+         elsif M = "in out" then
+            return Ada_Mode_In_Out;
+         else
+            return Ada_Mode_Default;
+         end if;
+      end Value;
+
+   begin
+      Target_Subp := Resolve_Name_Precisely (Get_Node_As_Name (Node));
+
+      if Target_Subp.Is_Null then
+         raise Program_Error with
```

Comment:
Can we set `Error` parameter instead of raising a `Program_Error` here?

Reply:
Agree.